File: android/app/src/main/kotlin/app/paddleup/services/SessionRecorder.kt
package app.paddleup.services

import app.paddleup.data.db.SessionDao
import app.paddleup.data.db.SessionEntity
import app.paddleup.data.db.TrackPointEntity
import app.paddleup.shared.geo.downsample
import app.paddleup.shared.model.CraftType
import app.paddleup.shared.model.SessionSource
import app.paddleup.shared.model.TrackPoint
import app.paddleup.shared.model.TrackSummaryPoint
import app.paddleup.shared.session.SessionAggregator
import com.google.firebase.auth.FirebaseAuth
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.plus
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

data class LiveStats(
    val isRecording: Boolean = false,
    val startedAtEpochMs: Long = 0,
    val durationSec: Double = 0.0,
    val distanceMeters: Double = 0.0,
    val currentSpeedMps: Double = 0.0,
    val currentHr: Int? = null,
    val currentStrokeRate: Double = 0.0,
    val strokeCount: Int = 0,
)

@Singleton
class SessionRecorder @Inject constructor(
    private val locationTracker: LocationTracker,
    private val motionTracker: MotionTracker,
    private val dao: SessionDao,
    private val auth: FirebaseAuth,
    private val json: Json,
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private var collectJobs: List<Job> = emptyList()

    private val _live = MutableStateFlow(LiveStats())
    val live: StateFlow<LiveStats> = _live.asStateFlow()

    private val track = mutableListOf<TrackPoint>()
    private var sessionId: String? = null
    private var startEpochMs: Long = 0
    private var lastStrokeRate: Double = 0.0
    private var strokeCount: Int = 0
    private var craftType: CraftType = CraftType.OC1

    fun start(craft: CraftType) {
        if (_live.value.isRecording) return
        craftType = craft
        sessionId = UUID.randomUUID().toString()
        startEpochMs = System.currentTimeMillis()
        track.clear()
        strokeCount = 0
        lastStrokeRate = 0.0
        _live.value = LiveStats(isRecording = true, startedAtEpochMs = startEpochMs)

        collectJobs = listOf(
            scope.launch {
                locationTracker.samples().collect { gps ->
                    val tSec = (gps.tEpochMs - startEpochMs) / 1000.0
                    val point = TrackPoint(
                        t = tSec,
                        lat = gps.lat,
                        lon = gps.lon,
                        altM = gps.altM,
                        speedMps = gps.speedMps,
                        hr = null,
                        strokeRate = if (lastStrokeRate > 0) lastStrokeRate else null,
                    )
                    track += point
                    val totals = SessionAggregator.totals(track, strokeCount)
                    _live.value = _live.value.copy(
                        durationSec = tSec,
                        distanceMeters = totals.distanceMeters,
                        currentSpeedMps = gps.speedMps,
                        currentStrokeRate = lastStrokeRate,
                        strokeCount = strokeCount,
                    )
                }
            },
            scope.launch {
                motionTracker.strokes().collect { stroke ->
                    strokeCount++
                    lastStrokeRate = stroke.rateSpm
                    _live.value = _live.value.copy(
                        currentStrokeRate = stroke.rateSpm,
                        strokeCount = strokeCount,
                    )
                }
            },
        )
    }

    suspend fun stopAndSave(): String? {
        if (!_live.value.isRecording) return null
        collectJobs.forEach { it.cancel() }
        collectJobs = emptyList()
        _live.value = _live.value.copy(isRecording = false)

        val id = sessionId ?: return null
        val uid = auth.currentUser?.uid ?: "anonymous"
        val totals = SessionAggregator.totals(track, strokeCount)
        val splits = SessionAggregator.splits(track)
        val hr = SessionAggregator.hrSummary(track)
        val summary = downsample(track, target = 200).map {
            TrackSummaryPoint(it.t, it.lat, it.lon, it.altM, it.speedMps)
        }

        val entity = SessionEntity(
            id = id,
            userId = uid,
            craftType = craftType.name,
            source = SessionSource.ANDROID_PHONE.name,
            appVersion = "0.1.0",
            startedAtEpochMs = startEpochMs,
            endedAtEpochMs = startEpochMs + (totals.durationSec * 1000).toLong(),
            distanceMeters = totals.distanceMeters,
            durationSec = totals.durationSec,
            avgPaceSecPerKm = totals.avgPaceSecPerKm,
            avgSpeedMps = totals.avgSpeedMps,
            maxSpeedMps = totals.maxSpeedMps,
            strokeCount = totals.strokeCount,
            avgStrokeRate = totals.avgStrokeRate,
            elevationGainM = totals.elevationGainM,
            avgHr = hr.avg,
            maxHr = hr.max,
            splitsJson = json.encodeToString(splits),
            hrZonesJson = json.encodeToString(hr.zones),
            sideSwitchesJson = "[]",
            weatherJson = null,
            trackSummaryJson = json.encodeToString(summary),
        )

        dao.upsertSession(entity)
        dao.insertTrackPoints(
            track.map {
                TrackPointEntity(
                    sessionId = id,
                    tSec = it.t,
                    lat = it.lat,
                    lon = it.lon,
                    altM = it.altM,
                    speedMps = it.speedMps,
                    hr = it.hr,
                    strokeRate = it.strokeRate,
                    cadenceConfidence = it.cadenceConfidence,
                )
            },
        )

        sessionId = null
        return id
    }

    fun discard() {
        collectJobs.forEach { it.cancel() }
        collectJobs = emptyList()
        track.clear()
        sessionId = null
        _live.value = LiveStats()
    }
}

File: ios/Paddleup/Services/SessionRecorder.swift
import Foundation
import PaddleupShared

@MainActor
final class SessionRecorder: ObservableObject {

    struct LiveStats {
        var isRecording: Bool = false
        var startedAt: Date = .distantPast
        var durationSec: Double = 0
        var distanceMeters: Double = 0
        var currentSpeedMps: Double = 0
        var currentHr: Int? = nil
        var currentStrokeRate: Double = 0
        var strokeCount: Int = 0
    }

    @Published private(set) var live = LiveStats()

    private let location = LocationTracker()
    private let motion = MotionTracker()
    private let store = SessionStore()
    var sessionStore: SessionStore { store }

    private var locationTask: Task<Void, Never>?
    private var motionTask: Task<Void, Never>?

    private var sessionId: String?
    private var craftType: CraftType = .OC1
    private var startedAt: Date = .distantPast
    private var track: [TrackPoint] = []
    private var strokeCount: Int = 0
    private var lastStrokeRate: Double = 0

    func start(craft: CraftType) {
        guard !live.isRecording else { return }
        self.craftType = craft
        self.sessionId = UUID().uuidString
        self.startedAt = Date()
        self.track.removeAll(keepingCapacity: true)
        self.strokeCount = 0
        self.lastStrokeRate = 0
        self.live = LiveStats(isRecording: true, startedAt: startedAt)

        locationTask = Task { [location, weak self] in
            for await s in location.stream() {
                guard let self else { return }
                let t = s.date.timeIntervalSince(self.startedAt)
                let p = TrackPoint(
                    t: t, lat: s.lat, lon: s.lon, altM: s.altM, speedMps: s.speedMps,
                    hr: nil,
                    strokeRate: self.lastStrokeRate > 0 ? self.lastStrokeRate : nil
                )
                self.track.append(p)
                let totals = SessionAggregator.totals(points: self.track, strokeCount: self.strokeCount)
                self.live.durationSec = t
                self.live.distanceMeters = totals.distanceMeters
                self.live.currentSpeedMps = s.speedMps
                self.live.strokeCount = self.strokeCount
            }
        }

        motionTask = Task { [motion, weak self] in
            for await s in motion.strokes() {
                guard let self else { return }
                self.strokeCount += 1
                self.lastStrokeRate = s.rateSpm
                self.live.currentStrokeRate = s.rateSpm
                self.live.strokeCount = self.strokeCount
            }
        }
    }

    func stopAndSave(uid: String) async -> Session? {
        guard live.isRecording, let id = sessionId else { return nil }
        locationTask?.cancel(); motionTask?.cancel()
        locationTask = nil; motionTask = nil

        let endedAt = Date()
        let totals = SessionAggregator.totals(points: track, strokeCount: strokeCount)
        let splits = SessionAggregator.splits(points: track)
        let hr = SessionAggregator.hrSummary(points: track)
        let summary = Geo.downsample(track, target: 200).map {
            TrackSummaryPoint(t: $0.t, lat: $0.lat, lon: $0.lon, altM: $0.altM, speedMps: $0.speedMps)
        }
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime]
        let session = Session(
            id: id,
            userId: uid,
            source: .iosPhone,
            appVersion: Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "0.1.0",
            craftType: craftType,
            startedAt: iso.string(from: startedAt),
            endedAt: iso.string(from: endedAt),
            totals: totals,
            hr: hr,
            splits: splits,
            sideSwitches: [],
            weather: nil,
            trackSummary: summary
        )
        try? store.save(session: session, track: track)
        live = LiveStats()
        sessionId = nil
        return session
    }

    func discard() {
        locationTask?.cancel(); motionTask?.cancel()
        locationTask = nil; motionTask = nil
        track.removeAll()
        sessionId = nil
        live = LiveStats()
    }
}
